### PR TITLE
Add nativemem test with dlopen after profiler start.

### PR DESCRIPTION
**Pull request by krk**
### Description
Parse symbols and hook functions of the main executable, when profiler is *not* loaded via `LD_PRELOAD`.

Without this fix, profiler is unable to hook non-java executables if `dlopen` is called after starting the profiler. 

### Related issues
Issue was introduced in https://github.com/async-profiler/async-profiler/pull/1220 as `dlerror` may *not* return `NULL` for the main executable.

Thanks for the repro in https://github.com/async-profiler/async-profiler/issues/1241 @Baraa-Hasheesh  and for discussions @fandreuz and @Baraa-Hasheesh.

Fixes https://github.com/async-profiler/async-profiler/issues/1233 and https://github.com/async-profiler/async-profiler/issues/1241.

### How has this been tested?
`make test-java TESTS=nativemem`

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: src/symbols_linux.cpp [hunk 1]
```diff
@@ -804,13 +817,20 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
...
-            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() returns NULL for them.
+            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() "in some cases" returns NULL for them.
             void* handle = dlopen(lib.file, RTLD_LAZY | RTLD_NOLOAD);
-            if (handle != NULL || dlerror() == NULL || OS::isMusl()) {
+            const char* dlerror_str = dlerror();
+
+            // Main executable will return NULL from handle, we need to parse it anyway.
+            // dlopen is not required in this case as the main exe cannot be unloaded.
+            bool isMainExe = handle == NULL && exePath != NULL && strcmp(lib.file, exePath) == 0;
+            bool dlopenSuccess = handle != NULL || dlerror_str == NULL || OS::isMusl();
```
Comment (fandreuz) [inline src/symbols_linux.cpp:827]: Maybe we don't need `OS::isMusl()` anymore? The only situation when this was appropriate is for the main executable, in which case `isMainExe` is `true` anyways

File: test/native/libs/callsmalloc.c [hunk 2]
```diff
@@ -0,0 +1,10 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+
+__attribute__((visibility("default"))) void* call_malloc(size_t size) {
```
Comment (fandreuz) [inline test/native/libs/callsmalloc.c:8]: Maybe you could reuse `Java_test_nativemem_Native_malloc` from `jnimalloc.c` instead of creating a new file?

File: test/one/profiler/test/RunnableTest.java [hunk 3]
```diff
@@ -25,7 +25,9 @@ public Test test() {
     }
 
     public String testName() {
-        return className() + '.' + m.getName();
+        return test.nameSuffix().isEmpty()
+            ? className() + '.' + m.getName()
+            : className() + '.' + m.getName() + ' ' + test.nameSuffix();
```
Comment (fandreuz) [inline test/one/profiler/test/RunnableTest.java:30]: return String.format("%s.%s %s", className(), m.getName(), test.nameSuffix()).trim();

File: test/test/nativemem/profile_with_dlopen.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+#include "asprof.h"
+#include <dlfcn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define ASSERT_NO_DLERROR()  \
+    err = dlerror();         \
+    if (err != NULL) {       \
+        printf("%s\n", err); \
```
Comment (fandreuz) [inline test/test/nativemem/profile_with_dlopen.c:15]: `stderr`? Same for L21

File: test/test/nativemem/profile_with_dlopen.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "asprof.h"
+#include <dlfcn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define ASSERT_NO_DLERROR()  \
```
Comment (fandreuz) [inline test/test/nativemem/profile_with_dlopen.c:12]: What's the advantage of `#define` over a function?

File: test/one/profiler/test/RunnableTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (apangin) [reply test/one/profiler/test/RunnableTest.java:30]: Kerem's version is better: it's more obvious and far more efficient (although the latter is not important here, let's not support bad practice).

File: src/symbols_linux.cpp [hunk 6]
```diff
@@ -775,6 +775,19 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
...
+        char buf[PATH_MAX];
+
+        // realpath() may fail for a path like /proc/[pid]/root/bin/asprof
+        // In this case, resolve the link as is.
+        ssize_t size = readlink("/proc/self/exe", buf, sizeof(buf) - 1);
+        if (size >= 0) {
+            buf[size] = 0;
+            exePath = buf;
+        }
+    }
```
Comment (apangin) [inline src/symbols_linux.cpp:789]: There is a simpler and more efficient way to find main executable:
```
    const void* main_phdr = NULL;
    dl_iterate_phdr([](struct dl_phdr_info* info, size_t size, void* data) {
        *(const void**)data = info->dlpi_phdr;
        return 1;
    }, &main_phdr);
```
Then check if `main_phdr >= lib.image_base && main_phdr < lib.map_end`

Calling `dl_iterate_phdr` should be fine as long as there are no allocations inside.

File: src/symbols_linux.cpp [hunk 7]
```diff
@@ -804,13 +817,20 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
...
             // Also, dlopen() ensures the library is fully loaded.
-            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() returns NULL for them.
+            // Main executable and ld-linux interpreter cannot be dlopen'ed, but dlerror() "in some cases" returns NULL for them.
             void* handle = dlopen(lib.file, RTLD_LAZY | RTLD_NOLOAD);
-            if (handle != NULL || dlerror() == NULL || OS::isMusl()) {
+            const char* dlerror_str = dlerror();
+
+            // Main executable will return NULL from handle, we need to parse it anyway.
+            // dlopen is not required in this case as the main exe cannot be unloaded.
+            bool isMainExe = handle == NULL && exePath != NULL && strcmp(lib.file, exePath) == 0;
```
Comment (apangin) [inline src/symbols_linux.cpp:826]: `snake_case` for variable names.

**Review by apangin — COMMENTED** [review]
Thank you for root causing the issue.

File: test/test/nativemem/NativememTests.java [hunk 8]
```diff
@@ -178,4 +178,14 @@ public void malloc_plt_dyn(TestProcess p) throws Exception {
         Assert.isEqual(sizeCounts.getOrDefault((long) MALLOC_SIZE, 0L), 1);
         Assert.isEqual(sizeCounts.getOrDefault((long) MALLOC_DYN_SIZE, 0L), 1);
     }
+
+    @Test(os = Os.LINUX, sh = "%testbin/profile_with_dlopen dlopen_first %f.jfr", output = true, env = {"LD_LIBRARY_PATH=build/test/lib:build/lib"}, nameSuffix = "dlopen_first")
+    @Test(os = Os.LINUX, sh = "%testbin/profile_with_dlopen profile_first %f.jfr", output = true, env = {"LD_LIBRARY_PATH=build/test/lib:build/lib"}, nameSuffix = "profile_first")
+    @Test(os = Os.LINUX, sh = "LD_PRELOAD=%lib %testbin/profile_with_dlopen dlopen_first %f.jfr", output = true, env = {"LD_LIBRARY_PATH=build/test/lib:build/lib"}, nameSuffix = "dlopen_first with LD_PRELOAD")
+    @Test(os = Os.LINUX, sh = "LD_PRELOAD=%lib %testbin/profile_with_dlopen profile_first %f.jfr", output = true, env = {"LD_LIBRARY_PATH=build/test/lib:build/lib"}, nameSuffix = "profile_first with LD_PRELOAD")
```
Comment (Baraa-Hasheesh) [inline test/test/nativemem/NativememTests.java:185]: We should add a new test that checks the dlopen patching on both MacOs & Linux?
As this issue of not patching the main executable is also relevant to MacOs

File: src/symbols_linux.cpp [hunk 9]
```diff
@@ -821,6 +841,8 @@ void Symbols::parseLibraries(CodeCacheArray* array, bool kernel_symbols) {
         array->add(cc);
     }
 
+    free(exePath);
```
Comment (Baraa-Hasheesh) [inline src/symbols_linux.cpp:844]: is this safe? as in some cases it points toward the `buf[PATH_MAX]` which is defined in the stack not the heap

File: test/test/nativemem/profile_with_dlopen.c [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        exit(1);             \
+    }
+
+#define ASSERT_NO_ASPROF_ERR(err)               \
+    if (err != NULL) {                          \
+        printf("%s\n", _asprof_error_str(err)); \
+        exit(1);                                \
+    }
+
+asprof_error_str_t _asprof_error_str;
```
Comment (Baraa-Hasheesh) [inline test/test/nativemem/profile_with_dlopen.c:25]: no need for this global variable, you can define it in the main function

**Comment by Baraa-Hasheesh** [comment]
@krk @apangin @fandreuz 

The above fix doesn't fully fix the issue of patching libs,
Due to how the dynamic linker work & common optimization it does -> if you do the following steps

```
void* lib1 = dlopen("libcallsmalloc.so", RTLD_NOW);
dlclose(lib1);
void2 lib2 = dlopen("libcallsmalloc.so", RTLD_NOW);
```

You will actually find that the `lib1` & `lib2` point to the same address even though the shared objects were deleted & recreated

=> you can try this in the new test by doing the following
image

This means due to the Async-profiler built in optimization of skipping already patched libs the second dlopen will never be patched as it will be seen as already patched from the first dlopen 

This is quite an old issue (Even the old logic before #1220 has the issue) that could require significant changes to address
* Hooking the dlclose calls & actually removing the patches from the async profiler cache if reference counts reaches 0 for a lib.
* We could need locking logic as it's possible race conditions may happen when dlclose-ing & dlopen-ing from multiple threads.

Interesting edge cases will be handling the `RTLD_NODELETE` flag as even if the reference counter reaches 0, there is no need to cleanup the async-profiler local cache as the shared objects are never unloaded.

File: test/test/nativemem/profile_with_dlopen.c (hunk 5 above)
Comment (krk) [reply test/test/nativemem/profile_with_dlopen.c:12]: We can exit directly, without checking the return value of a new function. I used macros here to reduce repetition.

File: test/test/nativemem/NativememTests.java (hunk 8 above)
Comment (krk) [reply test/test/nativemem/NativememTests.java:185]: Macos support is "best effort", will leave it out from the current PR.

File: src/symbols_linux.cpp (hunk 9 above)
Comment (krk) [reply src/symbols_linux.cpp:844]: Good catch! This will be deleted in the new version.
